File: clients/cpp/examples/subscribe.cpp
/**
 * @file subscribe.cpp
 * @brief Example: Subscribing to messages with NexusD C++ client
 * 
 * Demonstrates auto-reconnection with exponential backoff and gap detection.
 */

#include <nexusd_client/client.hpp>
#include <iostream>
#include <csignal>
#include <atomic>

std::atomic<bool> running{true};

void signalHandler(int) {
    std::cout << "\nShutting down...\n";
    running = false;
}

int main() {
    std::signal(SIGINT, signalHandler);
    
    try {
        // Configure reconnection policy
        nexusd::ReconnectionPolicy policy{
            .max_retries = 0,  // 0 = infinite retries
            .initial_delay_ms = 1000,
            .max_delay_ms = 30000,
            .exponential_backoff = true,
            .jitter_factor = 0.1,
        };

        // Connect to daemon with resilience settings
        nexusd::Client client("localhost:5672", policy);
        std::cout << "Connected to NexusD\n";
        std::cout << "Subscribing to sensors/* topics...\n";
        std::cout << "Auto-reconnect enabled with gap recovery\n";
        std::cout << "Press Ctrl+C to stop\n\n";

        // Register connection state callback
        client.setReconnectionCallback([](nexusd::ReconnectionState state, const std::string& error) {
            switch (state) {
                case nexusd::ReconnectionState::Connected:
                    std::cout << "[Connection] Connected to daemon\n";
                    break;
                case nexusd::ReconnectionState::Reconnecting:
                    std::cout << "[Connection] Reconnecting... (" << error << ")\n";
                    break;
                case nexusd::ReconnectionState::Disconnected:
                    std::cout << "[Connection] Disconnected\n";
                    break;
                case nexusd::ReconnectionState::Failed:
                    std::cout << "[Connection] Failed: " << error << "\n";
                    break;
            }
        });

        // Track last sequence number for gap detection
        uint64_t last_seq = 0;

        // Subscribe to multiple topics with gap recovery
        std::vector<std::string> topics = {
            "sensors/temperature",
            "sensors/humidity",
            "config/settings"
        };

        nexusd::SubscribeOptions opts{
            .receive_retained = true,
            .gap_recovery_mode = nexusd::GapRecoveryMode::ReplayBuffer,
        };

        client.subscribe(
            topics,
            // Message callback
            [&last_seq](const nexusd::Message& msg) {
                // Gap detection
                if (last_seq > 0 && msg.sequence_number > last_seq + 1) {
                    auto gap = msg.sequence_number - last_seq - 1;
                    std::cout << "[Gap] Detected " << gap << " missed messages on " << msg.topic << "\n";
                    std::cout << "       Expected seq " << (last_seq + 1) << ", got " << msg.sequence_number << "\n";
                }
                last_seq = msg.sequence_number;

                std::string retained_flag = msg.is_retained ? " [RETAINED]" : "";
                std::cout << "Topic: " << msg.topic << retained_flag << "\n";
                std::cout << "  Payload: " << msg.payloadAsString() << "\n";
                std::cout << "  Sequence: " << msg.sequence_number << "\n";
                std::cout << "  Message ID: " << msg.message_id << "\n";
                std::cout << "  Timestamp: " << msg.timestamp_ms << "\n";
                std::cout << "\n";
            },
            // Error callback
            [](const std::string& error) {
                std::cerr << "Subscription error: " << error << "\n";
            },
            opts
        );
        
        return 0;
        
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }
}

File: clients/rust/examples/subscribe.rs
//! Example: Subscribing to messages with NexusD Rust client
//! 
//! Demonstrates auto-reconnection with exponential backoff and gap detection.

use nexusd_client::{
    ReconnectingClient,
    ReconnectionPolicy,
    GapRecoveryMode,
    ReconnectionState,
};
use tokio_stream::StreamExt;
use tokio::signal;
use tracing::{info, warn, Level};
use tracing_subscriber;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Initialize tracing for structured logging
    tracing_subscriber::fmt()
        .with_max_level(Level::INFO)
        .init();

    // Configure reconnection policy
    let policy = ReconnectionPolicy {
        max_retries: None,  // None = infinite retries
        initial_delay: std::time::Duration::from_secs(1),
        max_delay: std::time::Duration::from_secs(30),
        exponential_backoff: true,
        jitter_factor: 0.1,
    };

    // Connect to daemon with resilience settings
    let mut client = ReconnectingClient::connect(
        "http://localhost:5672",
        policy,
    ).await?;
    
    info!("Connected to NexusD");
    println!("Subscribing to sensors/* topics...");
    println!("Auto-reconnect enabled with gap recovery");
    println!("Press Ctrl+C to stop\n");

    // Subscribe to multiple topics with gap recovery
    let topics = vec![
        "sensors/temperature".to_string(),
        "sensors/humidity".to_string(),
        "config/settings".to_string(),
    ];

    let mut stream = client.subscribe_with_reconnect(
        topics,
        GapRecoveryMode::ReplayBuffer,
    ).await?;

    // Process messages until Ctrl+C
    tokio::select! {
        _ = async {
            let mut last_seq: u64 = 0;
            
            while let Some(result) = stream.next().await {
                match result {
                    Ok(msg) => {
                        // Gap detection
                        if last_seq > 0 && msg.sequence_number > last_seq + 1 {
                            let gap = msg.sequence_number - last_seq - 1;
                            warn!(
                                topic = %msg.topic,
                                expected = last_seq + 1,
                                actual = msg.sequence_number,
                                gap = gap,
                                "Gap detected - missed {} messages",
                                gap
                            );
                        }
                        last_seq = msg.sequence_number;

                        let retained_flag = if msg.is_retained { " [RETAINED]" } else { "" };
                        println!("Topic: {}{}", msg.topic, retained_flag);

                        match msg.payload_as_string() {
                            Ok(payload) => println!("  Payload: {}", payload),
                            Err(_) => println!("  Payload: {:?}", msg.payload),
                        }

                        println!("  Sequence: {}", msg.sequence_number);
                        println!("  Message ID: {}", msg.message_id);
                        println!("  Timestamp: {}", msg.timestamp_ms);
                        println!();
                    }
                    Err(e) => {
                        warn!("Stream error: {}", e);
                    }
                }
            }
        } => {}
        _ = signal::ctrl_c() => {
            println!("\nShutting down...");
        }
    }

    Ok(())
}
